src/services/smart-fields-values-injector.js

```diff
@@ -17,6 +17,25 @@ function SmartFieldsValuesInjector(record, modelName, fieldsPerModel) {
       }
     }
 
+    // NOTICE: Update the record with smart field value (contains side effects).
+    async function updateRecord(smartFieldValue) {
+      // NOTICE: We need some recursion in order to consider smart fields in smart relations.
+      //         So if the smartFieldValue contains data and the field has a reference then we
+      //         consider we have to inject smart fields within the reference.
+      if (smartFieldValue && smartFieldValue.dataValues && field.reference) {
+        // NOTICE: Field reference format: `modelName.property`.
+        const [referenceModelName] = field.reference.split('.');
+        const smartFieldsValuesInjector = new SmartFieldsValuesInjector(
+          smartFieldValue,
+          referenceModelName,
+          fieldsPerModel,
+        );
+        await smartFieldsValuesInjector.perform();
+      }
+      // NOTICE: (side effects) Update the record, then update fieldsForHighlightedSearch.
```

Comment:
I removed every comments about side effects (see my previous comment). 

> If you think that addFieldForHighlightIfCandidate method name is not explicit enough, feel free to improve it 

I agree, it's not explicit. I removed it and replaced it by a comment:

```javascript
// NOTICE: String fields can be highlighted.
if (field.type === 'String') {
  fieldsForHighlightedSearch.push(field.field);
}
```

_Side note: there are still side effects in this function, I don't think I should to refactor it since I want to focus on the main issue._

Reply:
Really not a fan of 1 comment above each line of code. Like this:
```javascript
// NOTICE: Update the record with the the Smart Field value.
record[field.field] = smartFieldValue;
// NOTICE: String fields can be highlighted.
if (field.type === 'String') {
  fieldsForHighlightedSearch.push(field.field);
}
```
This is pretty limited so it does not hurt.
But comments are like code, you have to maintain them (and if you don't do it well, people that read them could be lost).
For instance, I am not sure we want to maintain:
```javascript
// NOTICE: Update the record with the the Smart Field value.
```
comment, as the intension described is really close to the action:
```javascript
record[field.field] = smartFieldValue;
```